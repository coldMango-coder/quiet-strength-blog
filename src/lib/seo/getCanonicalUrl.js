/**
 * Generates canonical URLs for trueallyguide.com pages
 * @param {string} pathname - The pathname from router (e.g., '/blog/post')
 * @returns {string} - The canonical URL
 */
export function getCanonicalUrl(pathname) {
  const baseUrl = 'https://trueallyguide.com';
  
  // Normalize pathname
  let normalizedPath = pathname || '/';
  
  // Remove query parameters and hash
  normalizedPath = normalizedPath.split('?')[0].split('#')[0];
  
  // Remove trailing slash except for root
  if (normalizedPath !== '/' && normalizedPath.endsWith('/')) {
    normalizedPath = normalizedPath.slice(0, -1);
  }
  
  // Convert to lowercase
  normalizedPath = normalizedPath.toLowerCase();
  
  // For root, return exact base URL
  if (normalizedPath === '/') {
    return `${baseUrl}/`;
  }
  
  return `${baseUrl}${normalizedPath}`;
}