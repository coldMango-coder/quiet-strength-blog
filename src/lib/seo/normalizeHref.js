/**
 * Normalizes internal links to ensure consistency with canonical URLs
 * @param {string} href - The href attribute to normalize
 * @returns {string} - The normalized href
 */
export function normalizeHref(href) {
  if (!href || typeof href !== 'string') {
    return href;
  }
  
  // Skip external links, mailto, tel, etc.
  if (href.startsWith('http') && !href.startsWith('http://trueallyguide.com') && !href.startsWith('https://trueallyguide.com') && !href.startsWith('http://www.trueallyguide.com') && !href.startsWith('https://www.trueallyguide.com')) {
    return href;
  }
  
  if (href.startsWith('mailto:') || href.startsWith('tel:') || href.startsWith('#') || href.startsWith('javascript:')) {
    return href;
  }
  
  let normalizedHref = href;
  
  // Convert http to https and remove www for our domain (case insensitive)
  normalizedHref = normalizedHref.replace(/^http:\/\/(www\.)?trueallyguide\.com/i, 'https://trueallyguide.com');
  normalizedHref = normalizedHref.replace(/^https:\/\/www\.trueallyguide\.com/i, 'https://trueallyguide.com');
  
  // Extract path for internal links
  let path = normalizedHref;
  if (normalizedHref.startsWith('https://trueallyguide.com')) {
    path = normalizedHref.replace('https://trueallyguide.com', '') || '/';
  }
  
  // Remove query parameters and hash for normalization (but keep in final URL)
  const [pathPart, queryAndHash] = path.split(/[?#]/);
  let normalizedPath = pathPart;
  
  // Convert to lowercase
  normalizedPath = normalizedPath.toLowerCase();
  
  // Remove trailing slash except for root
  if (normalizedPath !== '/' && normalizedPath.endsWith('/')) {
    normalizedPath = normalizedPath.slice(0, -1);
  }
  
  // Get query and hash parts from the original href
  const queryHashPart = href.includes('?') || href.includes('#') ? 
    href.substring(href.search(/[?#]/)) : '';
  
  // Return full URL if original was a full URL to our domain, otherwise return relative path
  if (href.toLowerCase().startsWith('http://') || href.toLowerCase().startsWith('https://')) {
    return `https://trueallyguide.com${normalizedPath === '/' ? '/' : normalizedPath}${queryHashPart}`;
  }
  
  // For relative paths, just return the normalized path
  return `${normalizedPath === '/' ? '/' : normalizedPath}${queryHashPart}`;
}