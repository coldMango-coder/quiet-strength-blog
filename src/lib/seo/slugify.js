import deburr from 'lodash.deburr';

// Transliterate to ASCII, lowercase, hyphenate, and strip non-url-safe chars
export default function slugify(input) {
  if (!input) return '';
  let s = String(input);
  // Remove diacritics using lodash.deburr
  s = deburr(s);
  // Lowercase
  s = s.toLowerCase();
  // Replace non-word sequences with hyphens
  s = s.replace(/[^a-z0-9]+/g, '-');
  // Trim hyphens
  s = s.replace(/^-+|-+$/g, '');
  return s;
}

