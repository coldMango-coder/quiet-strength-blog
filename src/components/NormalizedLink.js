import React from 'react';
import { Link } from 'react-router-dom';
import { normalizeHref } from '../lib/seo/normalizeHref';

/**
 * A wrapper around React Router's Link component that automatically normalizes hrefs
 * to ensure consistency with canonical URLs
 */
const NormalizedLink = ({ to, href, children, ...props }) => {
  // Support both 'to' (React Router) and 'href' (HTML anchor) props
  const originalHref = to || href;
  
  if (!originalHref) {
    return <a {...props}>{children}</a>;
  }
  
  const normalizedHref = normalizeHref(originalHref);
  
  // Handle null/undefined normalized href
  if (!normalizedHref || typeof normalizedHref !== 'string') {
    return <a {...props}>{children}</a>;
  }
  
  // Check if this is an external link or special link type
  const isExternal = normalizedHref.startsWith('http') && !normalizedHref.startsWith('https://trueallyguide.com');
  const isSpecial = normalizedHref.startsWith('mailto:') || 
                   normalizedHref.startsWith('tel:') || 
                   normalizedHref.startsWith('#') || 
                   normalizedHref.startsWith('javascript:');
  
  // Use regular anchor tag for external links and special links
  if (isExternal || isSpecial) {
    return (
      <a href={normalizedHref} {...props}>
        {children}
      </a>
    );
  }
  
  // Convert full URLs back to relative paths for React Router
  let routerPath = normalizedHref;
  if (normalizedHref.startsWith('https://trueallyguide.com')) {
    routerPath = normalizedHref.replace('https://trueallyguide.com', '') || '/';
  }
  
  // Use React Router Link for internal navigation
  return (
    <Link to={routerPath} {...props}>
      {children}
    </Link>
  );
};

export default NormalizedLink;