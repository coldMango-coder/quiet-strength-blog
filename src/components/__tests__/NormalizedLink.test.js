import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import { MemoryRouter } from 'react-router-dom';
import NormalizedLink from '../NormalizedLink';

// Use the actual normalizeHref function for more realistic testing
import { normalizeHref } from '../../lib/seo/normalizeHref';

const renderWithRouter = (component) => {
  return render(
    <MemoryRouter>
      {component}
    </MemoryRouter>
  );
};

describe('NormalizedLink', () => {
  test('renders internal links as React Router Links', () => {
    const { container } = renderWithRouter(
      <NormalizedLink to="/blog">Blog Link</NormalizedLink>
    );
    
    const link = container.querySelector('a');
    expect(link).toHaveTextContent('Blog Link');
    expect(link).toHaveAttribute('href', '/blog');
  });

  test('renders external links as anchor tags', () => {
    const { container } = renderWithRouter(
      <NormalizedLink to="https://external.com">External Link</NormalizedLink>
    );
    
    const link = container.querySelector('a');
    expect(link).toHaveTextContent('External Link');
    expect(link).toHaveAttribute('href', 'https://external.com');
  });

  test('renders special links as anchor tags', () => {
    const { container: mailtoContainer } = renderWithRouter(
      <NormalizedLink to="mailto:test@example.com">Email Link</NormalizedLink>
    );
    
    const { container: telContainer } = renderWithRouter(
      <NormalizedLink to="tel:+1234567890">Phone Link</NormalizedLink>
    );
    
    const { container: hashContainer } = renderWithRouter(
      <NormalizedLink to="#section">Hash Link</NormalizedLink>
    );
    
    expect(mailtoContainer.querySelector('a')).toHaveAttribute('href', 'mailto:test@example.com');
    expect(telContainer.querySelector('a')).toHaveAttribute('href', 'tel:+1234567890');
    expect(hashContainer.querySelector('a')).toHaveAttribute('href', '#section');
  });

  test('handles href prop as well as to prop', () => {
    const { container } = renderWithRouter(
      <NormalizedLink href="/about">About Link</NormalizedLink>
    );
    
    const link = container.querySelector('a');
    expect(link).toHaveTextContent('About Link');
    expect(link).toHaveAttribute('href', '/about');
  });

  test('handles missing href/to prop gracefully', () => {
    const { container } = renderWithRouter(
      <NormalizedLink>Empty Link</NormalizedLink>
    );
    
    const link = container.querySelector('a');
    expect(link).toHaveTextContent('Empty Link');
    expect(link).not.toHaveAttribute('href');
  });

  test('passes through additional props', () => {
    const { container } = renderWithRouter(
      <NormalizedLink to="/blog" className="custom-class" aria-label="Blog navigation">
        Blog Link
      </NormalizedLink>
    );
    
    const link = container.querySelector('a');
    expect(link).toHaveClass('custom-class');
    expect(link).toHaveAttribute('aria-label', 'Blog navigation');
  });

  test('normalizes internal trueallyguide.com URLs', () => {
    const { container } = renderWithRouter(
      <NormalizedLink to="https://trueallyguide.com/blog/">Internal URL</NormalizedLink>
    );
    
    const link = container.querySelector('a');
    expect(link).toHaveAttribute('href', '/blog');
  });
});

// Snapshot test for blog post content with internal links
describe('NormalizedLink Snapshot Test', () => {
  test('renders blog post content with normalized internal links', () => {
    // Simulate a blog post component with various types of links
    const BlogPostContent = () => (
      <div>
        <h1>Sample Blog Post</h1>
        <p>
          Check out our <NormalizedLink to="/Blog/">blog</NormalizedLink> for more articles.
        </p>
        <p>
          You might also like our article on{' '}
          <NormalizedLink to="/category/Self-Development/">self-development</NormalizedLink>.
        </p>
        <p>
          Read about <NormalizedLink to="https://trueallyguide.com/blog/introvert-tips/">introvert tips</NormalizedLink>.
        </p>
        <p>
          Visit our <NormalizedLink to="/">homepage</NormalizedLink> or{' '}
          <NormalizedLink href="mailto:contact@trueallyguide.com">contact us</NormalizedLink>.
        </p>
        <p>
          External reference: <NormalizedLink to="https://example.com">Example</NormalizedLink>
        </p>
        <p>
          Jump to <NormalizedLink to="#section">section</NormalizedLink> below.
        </p>
      </div>
    );

    const { container } = renderWithRouter(<BlogPostContent />);
    
    // Get all links and their normalized hrefs
    const links = container.querySelectorAll('a');
    const linkData = Array.from(links).map(link => ({
      text: link.textContent,
      href: link.getAttribute('href'),
      isRouterLink: !link.getAttribute('href').startsWith('http') && 
                    !link.getAttribute('href').startsWith('mailto:') && 
                    !link.getAttribute('href').startsWith('#')
    }));

    // Create snapshot of normalized links
    expect(linkData).toMatchSnapshot('normalized-internal-links');
    
    // Verify that internal links are normalized
    const internalLinks = linkData.filter(link => link.isRouterLink);
    expect(internalLinks).toEqual([
      { text: 'blog', href: '/blog', isRouterLink: true },
      { text: 'self-development', href: '/category/self-development', isRouterLink: true },
      { text: 'introvert tips', href: '/blog/introvert-tips', isRouterLink: true },
      { text: 'homepage', href: '/', isRouterLink: true }
    ]);

    // Verify external and special links are preserved
    const externalLinks = linkData.filter(link => !link.isRouterLink);
    expect(externalLinks).toEqual([
      { text: 'contact us', href: 'mailto:contact@trueallyguide.com', isRouterLink: false },
      { text: 'Example', href: 'https://example.com', isRouterLink: false },
      { text: 'section', href: '#section', isRouterLink: false }
    ]);
  });
});