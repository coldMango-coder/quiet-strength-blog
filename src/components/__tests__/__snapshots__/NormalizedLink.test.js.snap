// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NormalizedLink Snapshot Test renders blog post content with normalized internal links: normalized-internal-links 1`] = `
Array [
  Object {
    "href": "/blog",
    "isRouterLink": true,
    "text": "blog",
  },
  Object {
    "href": "/category/self-development",
    "isRouterLink": true,
    "text": "self-development",
  },
  Object {
    "href": "/blog/introvert-tips",
    "isRouterLink": true,
    "text": "introvert tips",
  },
  Object {
    "href": "/",
    "isRouterLink": true,
    "text": "homepage",
  },
  Object {
    "href": "mailto:contact@trueallyguide.com",
    "isRouterLink": false,
    "text": "contact us",
  },
  Object {
    "href": "https://example.com",
    "isRouterLink": false,
    "text": "Example",
  },
  Object {
    "href": "#section",
    "isRouterLink": false,
    "text": "section",
  },
]
`;
